### PR TITLE
has_function: Improve builtin detection

### PR DESCRIPTION
**Pull request by nirbheek**
* has_function: Improve built-in detection with headers

  - __has_builtin(foo) is more correct and more reliable than __has_builtin(__builtin_foo)
  - If detection of built-ins with headers fails in the first check because we can't take the address of a built-in, we can do an ordinary check.

We can't always do this ordinary check because clang on macOS ld removes the symbol when using -Wl,-no_weak_imports, which is needed for being able to target the correct platform with XCode 8 and newer.

Closes https://github.com/mesonbuild/meson/issues/3672

* has_function: Fix checking for __builtin_foo functions

Some functions are meant to be used only as compiler builtins, such as `__builtin_smul_overflow` and friends.

Detect this case and don't look for `__builtin_foo` because such a check will never include a header.

Fixes https://github.com/mesonbuild/meson/issues/3676

### CODE REVIEW

**Comment by textshell** [comment]
I think this wrongly attributes the problem to MinGW when this just reveals less than universally true a assumptions in meson. The current code also fails on linux where gcc is one of the two native compilers in #3676.

Meson should not assume a fixed relationship between `foo` and `__builtin_foo`. If there is a relationship is has to come from a system header.

This of course is breaking for projects where the build file failed to specify the needed header. We have to solve that in some way before we can do the proper fix.

**Comment by nirbheek** [comment]
> I think this wrongly attributes the problem to MinGW

The current commits do not do this.

> The current code also fails on linux where gcc is one of the two native compilers in #3676.

Now contains a fix for that issue + a test with a note about gcc being broken.

> This of course is breaking for projects where the build file failed to specify the needed header. We have to solve that in some way before we can do the proper fix.

Do tell us if you find a way to fix this. We have no choice but to do this for now.

**Comment by textshell** [comment]
I think for compatibility we need to deprecate this implicit assumption. Ideally meson would print a warning, but work as previously for a few releases. One way to do this would be to collect a list of affected functions and their needed includes and just implicitly add that include for the deprecation period.

From a grep of my include directory it seems like these functions might be affected.
(Using ```egrep -r '#.*__builtin_' /usr/include/``` on a debian stretch, but with newlib and diet libc headers also installed)

fpclassify isfinite isnormal isgreater isgreaterequal isless islessequal islessgreater isunordered alloca 
isnan va_start va_end va_arg offsetof strncpy strncat strcspn strspn strpbrk                  

It also reveals that the naming relation is often a bit different so the old code already does not work reliable, e.g.:
```
#define    __bswap16(_x)   __builtin_bswap16(_x)
#define isinf(__x) (__builtin_isinf_sign (__x))
#define __unreachable() __builtin_unreachable()
```

That list seems manageable although it is a bit long.

File: mesonbuild/compilers/c.py [hunk 1]
```diff
@@ -678,7 +678,8 @@ def has_function(self, funcname, prefix, env, extra_args=None, dependencies=None
             head, main = self._no_prototype_templ()
         templ = head + stubs_fail + main
 
-        if self.links(templ.format(**fargs), env, extra_args, dependencies):
+        if not funcname.startswith('__builtin_') and \
```
Comment (textshell) [inline mesonbuild/compilers/c.py]: I do not thing meson should have this assumption, why is this needed? Optimization?

File: mesonbuild/compilers/c.py [hunk 2]
```diff
@@ -692,10 +693,11 @@ def has_function(self, funcname, prefix, env, extra_args=None, dependencies=None
         # need to look for them differently. On nice compilers like clang, we
         # can just directly use the __has_builtin() macro.
         fargs['no_includes'] = '#include' not in prefix
+        fargs['func_is_builtin'] = funcname.startswith('__builtin_')
         t = '''{prefix}
         int main() {{
         #ifdef __has_builtin
-            #if !__has_builtin(__builtin_{func})
+            #if !__has_builtin({func})
```
Comment (textshell) [inline mesonbuild/compilers/c.py]: :+1:

But does this work without also adding implicit includes for the cases where users currently got away with not giving the includes?

File: mesonbuild/compilers/c.py [hunk 3]
```diff
@@ -704,10 +706,12 @@ def has_function(self, funcname, prefix, env, extra_args=None, dependencies=None
              * We would always check for this, but we get false positives on
              * MSYS2 if we do. Their toolchain is broken, but we can at least
              * give them a workaround. */
-            #if {no_includes:d}
+            #if {no_includes:d} && !{func_is_builtin:d}
                 __builtin_{func};
+            #elif defined(__stub_{func}) || defined(_stub__{func})
```
Comment (textshell) [inline mesonbuild/compilers/c.py]: Do we need a comment similar to the one near stubs_fail to make it easier to understand. 
Do we need to make sure that limits.h is included? Because the comment above implies it is needed for this to work.

File: test cases/common/43 has function/meson.build [hunk 4]
```diff
@@ -71,10 +73,10 @@ foreach cc : compilers
            args : unit_test_args),
            'built-in alloca must be found with #include')
     if not cc.compiles(defines_has_builtin, args : unit_test_args)
-      assert(not cc.has_function('alloca',
+      assert(cc.has_function('alloca',
              prefix : '#include <alloca.h>\n#undef alloca',
              args : unit_test_args),
-             'built-in alloca must not be found with #include and #undef')
+             'built-in alloca must be found with #include and #undef')
```
Comment (textshell) [inline test cases/common/43 has function/meson.build]: How does this work? It seems like it should not?

File: test cases/common/43 has function/meson.build [hunk 5]
```diff
@@ -88,4 +90,22 @@ foreach cc : compilers
...
+  if cc.has_function('strndup')
+    assert(cc.has_function('strndup', prefix: '#include <string.h>'), 'strndup not found with prefix')
+    assert(cc.links('#include <string.h>\nint main() { char *s = strndup ("string", 1); }'), 'strndup incorrectly detected')
+  endif
+
+  # This is meant to be used as __builtin_smul_overflow, so we shouldn't find the unbuiltin version
+  if ['gcc', 'clang'].contains(cc.get_id())
+    assert(cc.has_function('__builtin_smul_overflow'), '__builtin_smul_overflow must be found')
+    # FIXME: this fails on gcc, and we can't fix it without breaking builtin
+    # detection without headers for things like alloca()
```
Comment (textshell) [inline test cases/common/43 has function/meson.build]: I think with the implicit include workaround we can enable this test for gcc as well.

File: mesonbuild/compilers/c.py [hunk 6]
```diff
@@ -704,10 +706,12 @@ def has_function(self, funcname, prefix, env, extra_args=None, dependencies=None
              * We would always check for this, but we get false positives on
              * MSYS2 if we do. Their toolchain is broken, but we can at least
              * give them a workaround. */
-            #if {no_includes:d}
+            #if {no_includes:d} && !{func_is_builtin:d}
```
Comment (textshell) [inline mesonbuild/compilers/c.py]: I think we should try if the idea with implicit includes allows us to avoid this.

File: mesonbuild/compilers/c.py (hunk 2 above)
Comment (nirbheek) [reply mesonbuild/compilers/c.py]: Yes, the alloca test checks for this already (on clang).

File: mesonbuild/compilers/c.py (hunk 1 above)
Comment (nirbheek) [reply mesonbuild/compilers/c.py]: Yes, that check will always fail with builtins because it tries to take the address of the function.

Comment (textshell) [reply mesonbuild/compilers/c.py]: Which is fine, it just fails and meson goes on to the builtin test.
But what happens if there is a function that matches that pattern that is not a builtin? I think hardcoding that naming convention is just assuming things that might not be true. So i would always let that test fail. And if in some odd environment a function with such an name is an actual external function it will not break just because of naming.

File: test cases/common/43 has function/meson.build (hunk 4 above)
Comment (nirbheek) [reply test cases/common/43 has function/meson.build]: The test was added to ensure that the `__builtin_foo` codepath is not hit when there is a `#include` in prefix because there was a `#error` in the other path. That other path is gone now, and we have a fallback check for `foo` there now.

`alloca` may not be a `#define` now, but it is still available as a symbol. See: `alloca.h`.

File: mesonbuild/compilers/c.py (hunk 3 above)
Comment (nirbheek) [reply mesonbuild/compilers/c.py]: Fixed.

File: mesonbuild/compilers/c.py (hunk 6 above)
Comment (nirbheek) [reply mesonbuild/compilers/c.py]: I think we should do that in a separate PR since it has a large potential for a yak-shave.

File: mesonbuild/compilers/c.py (hunk 2 above)
Comment (textshell) [reply mesonbuild/compilers/c.py]: Yes, it seems to work with `alloca`, but with `isgreaterequal` it does not. Now it might be my clang install (it's just for compiling qtcreator). But a simple manual build test with clang seems to work. But even `has_function('isgreaterequal', prefix: '#include <math.h>')` does not work for me at the moment. It works with a random older meson.

File: mesonbuild/compilers/c.py (hunk 6 above)
Comment (textshell) [reply mesonbuild/compilers/c.py]: If this is the only place where we would need that, i'm ok with splitting that part of into a separate PR.

**Comment by textshell** [comment]
Sorry, not yet gotten to review the current version.

But from discussing with @ePirat and from @ePirat's testing i stumbled over this "nice" loophole with builtins:

https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html says:

> With the exception of built-ins that have library equivalents such as the standard C library functions discussed below, or that expand to library calls, GCC built-in functions are always expanded inline and thus do not have corresponding entry points and their address cannot be obtained. Attempting to use them in an expression other than a function call results in a compile-time error.
> 
> GCC includes built-in versions of many of the functions in the standard C library. These functions come in two forms: one whose names start with the __builtin_ prefix, and the other without. Both forms have the same type (including prototype), the same address (when their address is taken), and the same meaning as the C library functions even if you specify the -fno-builtin option see C Dialect Options). Many of these functions are only optimized in certain cases; if they are not optimized in a particular case, a call to the library function is emitted.
> 

And then goes on to list a ton of functions that are special in this way. So all functions from the list are builtins and still need their implementation to be available for linking. So this basically makes this a lot harder to do right.

But i think passing `-fno-builtin` to the test compilation would help. Btw, that option name is a bit misleading, as it's documented to only disable builtins that do not start with `__builtin`. But i would expect it to also not disable the "mandatory" builtins that don't have a non builtin version, but that would need checking.

Testing by @ePirat suggests that this would for `nstrdup` as well as the cases covered by the existing tests.

**Comment by nirbheek** [comment]
I didn't get enough time to look into this, and it's too dangerous to change this now because it might regress existing users. Will work on this after the 0.47 release which will happen today.

**Comment by nirbheek** [comment]
This is WIP, not ready to merge yet. I have to add the 'loophole' that @textshell pointed out above.

**Comment by ePirat** [comment]
#3902 is somewhat related

**Comment by nirbheek** [comment]
Fixed by https://github.com/mesonbuild/meson/pull/7116.
